packages/upset/src/provenance/index.ts

```diff
@@ -1,17 +1,43 @@
 /* eslint-disable @typescript-eslint/explicit-function-return-type */
 import {
-  AggregateBy, Plot, PlotInformation, SortByOrder, SortVisibleBy, UpsetConfig, DefaultConfig, Row, AltText,
+  AggregateBy, Plot, PlotInformation, SortByOrder, SortVisibleBy, UpsetConfig, DefaultConfig, Row,
+  Bookmark, BookmarkedSelection,
+  convertConfig,
+  ColumnName,
+  AltText,
 } from '@visdesignlab/upset2-core';
 
-import { Registry, initializeTrrack } from '@trrack/core';
+import { Registry, StateChangeFunction, initializeTrrack } from '@trrack/core';
 
 export type Metadata = {
   [key: string]: unknown;
 };
 
 const registry = Registry.create();
 
-const firstAggAction = registry.register(
+/**
+ * Registers a new action that uses a StateChangeFunction with the provenance registry while also guaranteeing
+ * that old upset config types (from outdated Trrack graph imports)
+ * are converted to the new upset config type before being passed to the StateChangeFunction.
+ * One type parameter is required; for the payload argument received by the action function.
+ * @param type Action type, string
+ * @param func Action function
+ * @typeparam DoActionPayload The type of the payload argument received by the action function (required)
+ * @typeparam UndoActionType The type of the undo action (optional)
+ * @typeparam UndoActionPayload The type of the payload argument received by the undo action function (optional)
+ * @returns An action creator that can be passed to provenance.apply
+ */
+function register<DoActionPayload, UndoActionType extends string = string, UndoActionPayload = any>(
+  type: string,
+  func: StateChangeFunction<UpsetConfig, DoActionPayload>,
+): ReturnType<typeof registry.register<typeof type, UndoActionType, DoActionPayload, UndoActionPayload, UpsetConfig>> {
```

Comment:
Is this trivially inferred by typescript?
```suggestion
) {
```

Reply:
Yes, I wanted to write down the type for clarity but it's very complex and I ended up missing the forest for the trees here; removed.